src/etlng/impl/Extraction.cpp

```diff
@@ -0,0 +1,227 @@
+        case PBObjType::CREATED:
+            return model::Object::ModType::Created;
+        case PBObjType::MODIFIED:
+            return model::Object::ModType::Modified;
+        case PBObjType::DELETED:
+            return model::Object::ModType::Deleted;
+        default:  // some gRPC system values that we don't care about
+            ASSERT(false, "Tried to extract bogus mod type '{}'", PBObjType::ModificationType_Name(type));
+    }
+
+    std::unreachable();
+}
+
+model::Transaction
+extractTx(PBTxType tx, uint32_t seq)
+{
+    auto raw = std::move(*tx.mutable_transaction_blob());
+    ripple::SerialIter it{raw.data(), raw.size()};
+    ripple::STTx const sttx{it};
+    ripple::TxMeta meta{sttx.getTransactionID(), seq, tx.metadata_blob()};
+
+    static constexpr std::size_t KeySize = 32;
+    std::string keyStr{reinterpret_cast<char const*>(sttx.getTransactionID().data()), KeySize};
+
+    return {
+        .raw = std::move(raw),
+        .metaRaw = std::move(*tx.mutable_metadata_blob()),
+        .sttx = sttx,  // trivially copyable
+        .meta = std::move(meta),
+        .id = sttx.getTransactionID(),
+        .key = std::move(keyStr),
+        .type = sttx.getTxnType()
+    };
+}
+
+std::vector<model::Transaction>
+extractTxs(PBTxListType transactions, uint32_t seq)
+{
+    namespace rg = std::ranges;
+    namespace vs = std::views;
+
+    // TODO: should be simplified with ranges::to<> when available
+    std::vector<model::Transaction> output;
+    output.reserve(transactions.size());
+
+    rg::move(transactions | vs::transform([seq](auto&& tx) { return extractTx(tx, seq); }), std::back_inserter(output));
+    return output;
+}
+
+model::Object
+extractObj(PBObjType obj)
+{
+    auto key = ripple::uint256::fromVoidChecked(obj.key());
+    ASSERT(key.has_value(), "Failed to deserialize key from void");
+
+    auto value_or = [](std::string const& maybe, std::string fallback) -> std::string {
+        if (maybe.empty())
+            return fallback;
+        return maybe;
+    };
+
+    return {
+        .key = *key,  // trivially copyable
+        .keyRaw = std::move(*obj.mutable_key()),
+        .data = {obj.mutable_data()->begin(), obj.mutable_data()->end()},
+        .dataRaw = std::move(*obj.mutable_data()),
+        .successor = value_or(obj.successor(), uint256ToString(data::firstKey)),
+        .predecessor = value_or(obj.predecessor(), uint256ToString(data::lastKey)),
+        .type = extractModType(obj.mod_type()),
+    };
+}
+
+std::vector<model::Object>
+extractObjs(PBObjListType objects)
+{
+    namespace rg = std::ranges;
+    namespace vs = std::views;
+
+    // TODO: should be simplified with ranges::to<> when available
+    std::vector<model::Object> output;
+    output.reserve(objects.size());
+
+    rg::move(objects | vs::transform([](auto&& obj) { return extractObj(obj); }), std::back_inserter(output));
+    return output;
+}
+
+model::BookSuccessor
+extractSuccessor(PBBookSuccessorType successor)
```

Comment:
Do we want to move from successor?

Reply:
I think we can. Notice how it's a copy in the arg, not a ref of some sort